heirloom/date/date.1
'\" t
.\" Copyright (c) 2003 Gunnar Ritter
.\"
.\" This software is provided 'as-is', without any express or implied
.\" warranty. In no event will the authors be held liable for any damages
.\" arising from the use of this software.
.\"
.\" Permission is granted to anyone to use this software for any purpose,
.\" including commercial applications, and to alter it and redistribute
.\" it freely, subject to the following restrictions:
.\"
.\" 1. The origin of this software must not be misrepresented; you must not
.\"    claim that you wrote the original software. If you use this software
.\"    in a product, an acknowledgment in the product documentation would be
.\"    appreciated but is not required.
.\"
.\" 2. Altered source versions must be plainly marked as such, and must not be
.\"    misrepresented as being the original software.
.\"
.\" 3. This notice may not be removed or altered from any source distribution.
.\" Sccsid @(#)date.1	1.14 (gritter) 10/11/03
.TH DATE 1 "10/11/03" "Heirloom Toolchest" "User Commands"
.SH NAME
date \- print or set the date
.SH SYNOPSIS
\fBdate\fR [\fB\-u\fR] [\fB+\fIformat\fR]
.br
\fBdate\fR [\fB\-u\fR] \fIHHMM\fR | \fImmddHHMM\fR[[\fIcc\fR]\fIyy\fR]
.br
\fBdate\fR \fB\-a\fR [\fB\-\fR]\fIsss.fff\fR
.SH DESCRIPTION
.B
Date
displays the current date if invoked without arguments
or with a
.BI + format
argument.
.PP
If the first argument does not start with
.BR + ,
.B date
sets the system time to the given value,
which may consist of
.sp
.RS
.TS
lfB l.
mm	month of year (01\^\(en\^12)
dd	day of month (01\^\(en\^31)
HH	hour (00\^\(en\^23)
MM	minute (00\^\(en\^59)
cc	first two digits of year
yy	second two digits of year
.TE
.RE
.PP
The following options alter the behavior of
.BR date :
.TP
\fB\-a\fR [\fI\-\fR]\fIsss.fff\fR
Adjust the system clock by
.I sss
seconds and
.I fff
milliseconds,
speeding up or,
with
.IR \-,
slowing down.
.TP
.B \-u
Use
.SM UTC
values for printing and setting the date.
.PP
An argument starting with
.B +
specifies a format string to print the current date.
A percent character
.B %
introduces a field descriptor with one or two following characters
which is replaced by a part of the current date,
as shown in the table below.
Other characters are printed literally.
A newline character is printed at the end of the output.
.TP
Field descriptors:
.TS
l3 l.
a	T{
locale-specific abbreviated weekday name (System\ III)
T}
.\" the following line implies the minimum table width
A	locale-specific full weekday name (SVID3, POSIX.2)
b	T{
locale-specific abbreviated month name (SVID3, POSIX.2)
T}
B	T{
locale-specific full month name (SVID3, POSIX.2)
T}
c	T{
locale-specific date and time (SVID3, POSIX.2)
T}
C	T{
century \(en 19 or 20 (POSIX.2) (\fB/usr/5bin/s42/date\fR, \fB/usr/5bin/posix/date\fR)
.br
default output format (\fB/usr/5bin/date\fR)
T}
d	T{
day of month \(en 01 to 31 (System\ III)
T}
D	T{
date as mm/dd/yy (System\ III)
T}
e	T{
day of month \(en 1 to 31 (SVID3, POSIX.2)
T}
F	T{
date as yyyy\-mm\-dd
T}
g	T{
week-based year within century \(en 00 to 99
T}
G	T{
week-based year including century \(en 0000 to 9999
T}
h	T{
same as %b (System\ III)
T}
H	T{
hour \(en 00 to 23 (System\ III)
T}
I	T{
hour \(en 01 to 12 (SVID3, POSIX.2)
T}
j	T{
day of year \(en 001 to 366 (System\ III)
T}
m	T{
month of year \(en 01 to 12 (System\ III)
T}
M	T{
minute \(en 00 to 59 (System\ III)
T}
n	T{
newline character (System\ III)
T}
N	T{
default output format
T}
p	T{
locale-specific AM\^/\^PM string (SVID3, POSIX.2)
T}
r	T{
locale-specific time using AM\^/\^PM notation (System\ III)
T}
R	T{
same as %H:%M (SVID3)
T}
S	T{
second \(en 00 to 61 (System\ III)
T}
t	T{
tab character (System\ III)
T}
T	T{
time as HH:MM:SS (System\ III)
T}
u	T{
weekday \(en 1 (Monday) to 7 (POSIX.2)
T}
U	T{
week of year (Sunday as first day of week) \(en 00 to 53 (SVID3, POSIX.2)
T}
V	T{
week of year (Monday as first day of week) \(en 01 to 53 (POSIX.2)
T}
w	T{
weekday \(en 0 (Sunday) to 6 (System\ III)
T}
W	T{
week of year (Monday as first day of week) \(en 00 to 53 (SVID3, POSIX.2)
T}
x	T{
locale-specific date (POSIX.2)
T}
X	T{
locale-specific time (SVID3, POSIX.2)
T}
y	T{
last two digits of year \(en 00 to 99 (System\ III)
T}
Y	T{
four-digit year \(en 1970 to 2037 (SVID3, POSIX.2)
T}
z	T{
numeric timezone
T}
Z	T{
timezone name (SVID3, POSIX.2)
T}
%	T{
percent sign (System\ III)
T}
Ec	T{
alternative date and time (POSIX.2)
T}
EC	T{
alternative base year (period) (POSIX.2)
T}
Ex	T{
alternative date (POSIX.2)
T}
EX	T{
alternative time (SUSv2)
T}
Ey	T{
offset from %EC in years (POSIX.2)
T}
EY	T{
full alternative year (POSIX.2)
T}
Od	T{
day of month filled with leading zero using alternative numerals (POSIX.2)
T}
Oe	T{
day of month filled with leading space using alternative numerals (POSIX.2)
T}
OH	T{
24-hour clok using alternative numerals (POSIX.2)
T}
OI	T{
12-hour clock using alternative numerals (POSIX.2)
T}
Om	T{
month of year using alternative numerals (POSIX.2)
T}
OM	T{
minute using alternative numerals (POSIX.2)
T}
OS	T{
second using alternative numerals (POSIX.2)
T}
Ou	T{
weekday (Monday = 1) using alternative numerals (POSIX.2)
T}
OU	T{
week of year (cf. %U) using alternative numerals (POSIX.2)
T}
OV	T{
week of year (cf. %V) using alternative numerals (POSIX.2)
T}
Ow	T{
weekday (Sunday = 0) using alternative numerals (POSIX.2)
T}
OW	T{
week of year (cf. %W) using alternative numerals (POSIX.2)
T}
Oy	T{
year (offset from %C) using alternative numerals (POSIX.2)
T}
.TE
.TP
Compatibility references:
.TS
l l.
.\" the following line implies the minimum table width
System III	AT&T System III manual page date(1).\ \ \ \ \ \ 
SVID3	T{
AT&T System V Interface Definition, Third Edition, Volume II, 1992,
date(BU_CMD). Includes all field descriptors present in System III.
T}
POSIX.2	T{
ISO/IEC 9945-2:1993(E), 4.15.4.1\^\(en\^4.15.4.2, pp. 241\^\(en\^242.
Includes all field descriptors present in System III.
T}
SUSv2	T{
The Open Group,
Single \s-1UNIX\s0
Specification, Version 2, \(co 1997, date(XCU).
Includes all field descriptors present in POSIX.2.
T}
.TE
.PP
When no arguments are given,
the format defaults to
.RS
.sp
%a %b %e %H:%M:%S %Z %Y
.sp
.RE
in the C or POSIX locale, resulting in an output like
.RS
.sp
Wed Feb 26 17:48:54 CET 2003
.sp
.RE
.SH "ENVIRONMENT VARIABLES"
.TP
.B "LANG, LC_ALL"
See
.IR locale (7).
.TP
.B LC_TIME
Alters the strings printed when format arguments are given.
.TP
.B TZ
Timezone; if unset, the system default is used.
.SH "SEE ALSO"
adjtime(2),
stime(2),
strftime(3),
locale(7)
.SH NOTES
After setting the date,
the
.IR hwclock (8)
utility is called to adjust the hardware clock on Linux systems.


ex-vi/Changes
Release ...
* The screen buffers for visual mode are now dynamically allocated, so
  vi usually does not return to ex mode with "screen too large" when the
  terminal is resized on a large monitor anymore.
* ex can now edit files with lines of arbitrary length. vi currently only
  shows the beginning of a line that does not fit onto the screen in its
  entirety.
* Viewing executables and compressed files is no longer inhibited.

Release 3/25/05
* vi no longer dies with a segmentation fault if a line does not fit on the
  screen after an insertion.
* The 'p' command now works correctly if the buffer contains a partial line
  with multibyte characters.
* Traditional regular expressions sometimes failed to operate correctly
  since the last release.

Release 2/25/05
* Traditional regular expressions can now be used with multibyte characters.
* When the 'ignorecase' option is toggled, saved regular expressions are now
  updated accordingly. (P)
* If a line began with a tabulator and another tabulator was inserted with
  the cursor located on the first tabulator, the display was not updated
  appropriately since the last release (Bugreport by Matthew Fischer). (P)
* Fixed a segmentation fault that occured in multibyte locales when operation
  was continued after vi had been stopped by ^Z, with the cursor positioned
  in a line longer than the terminal width as the last line of the current
  screen.
* Made multicolumn characters work in lines longer than the screen width in
  visual mode (Bugreport by Matthew Fischer).
* Made it work for Big5 locales (Patches by Matthew Fischer).
* Fixed a problem with the 'r' command in EUC-JP and Big5 locales (Bugreport
  by Matthew Fischer).
* The insertion of multicolumn characters and tab characters in multibyte
  locales now works with terminals that have the 'ic' but no 'im' termcap
  capability (Bugreport by Matthew Fischer).
* The argument to the -w option is correctly recognized now.
* If the SHELL environment variable is set to the empty string, it is now
  ignored.
* A non-null exit status is now returned if a file could not be opened, if
  an invalid address is given for a command, or if a tag is not found.
* If the match for a substitution is of length zero, a line overflow is
  now avoided when compiled with the 'UNIX(R) Regular Expression Library'.
* When moving left while the cursor is positioned over a multicolumn
  character at the end of the line, the bell is rung now (Bugreport by
  Matthew Fischer).
* When moving up or down to a row with different column arrangement while
  the cursor is positioned over a multicolumn character, the leftmost
  character above the original position is chosen in the new row.
* If both the -t and the -c option are given, the -t option is now
  processed first, i.e. the command is executed at the position where
  the tag was found.
* The -w option now also sets the scroll size for the 'z' command.
* When the name of a nonexisting file is given with the 'edit' ex command,
  the command now succeeds and prints a "[New file]" message.
* If standard output is not a terminal, no '\r' is written at the end of
  printed lines anymore.
* The 'source' ex command now works if command input comes from a pipe or
  regular file.
* Ex does not exit on errors immediately anymore if standard input is not
  a terminal but a pipe or regular file.
* The 'substitute' ex command can now be abbreviated as 'sub', 'subst' etc.
* A new version of mapmalloc.c that is derived from Unix 7th Edition code
  has been introduced.
* If the 'next!' ex command is given and the autowrite option is set, the
  current file is not written anymore.

Release 1/19/05
* The last release erroneously made 'X' work like 'x' in visual mode. It now
  deletes the character before the cursor again as documented (Bugreport by
  Matthew Fischer). (P)
* When a multicolumn character was replaced by another multicolumn character
  in insert mode, the display was not updated appropriately with terminals
  other than xterm.
* Various rendering errors happened with multicolumn characters if they
  started at an even column (counting from 1 upwards).
* When a multicolumn character was inserted and then replaced, the visual
  screen representation was sometimes not updated accordingly.
* Undoing the replacement of a multicolumn character by a singlecolumn
  character twice made the singlecolumn character invisible.
* The 'cw' visual command with a multibyte character as last character of
  the affected word located at the end of the line left garbage bytes past
  the end of the new text.
* Visual 'U' followed by 'u' lead to garbage on the screen when multibyte
  characters were present on the changed line.
* The position of '$' when changing text was fixed for cases where the first
  changed character had multiple columns but the last one had not.
* The handling of multicolumn characters was fixed for terminals without the
  IM (insert mode) capability. It is unlikely that such terminals actually
  exist, but vi will use the corresponding code for open mode if a termcap
  entry is not available.
* When an illegal multibyte sequence is entered in vi insert mode, no garbage
  bytes are generated anymore when the insert mode is left.
* The '\u', '\l', '\U', and '\L' substitution sequences work with multibyte
  characters now.
* Handle character case conversions with the '~' vi command correctly if the
  length of the converted multibyte sequence is smaller than the original one.
* Multibyte sequences that correspond to an unprintable character are now
  printed as multiple octal escape sequences.
* Pressing the ^U (kill) character in insert mode with a multibyte character
  at the beginning of an insertion at the start of a line sometimes positioned
  the cursor at weird locations since the last revision.
* Fixed an old vi bug: If a vi command that yanked or deleted part of a line
  was followed by an ex command that also yanked or deleted some text, a

heirloom/date/date.c
/*
 * date - write the date and time
 *
 * Gunnar Ritter, Freiburg i. Br., Germany, June 2002.
 */
/*
 * Copyright (c) 2003 Gunnar Ritter
 *
 * This software is provided 'as-is', without any express or implied
 * warranty. In no event will the authors be held liable for any damages
 * arising from the use of this software.
 *
 * Permission is granted to anyone to use this software for any purpose,
 * including commercial applications, and to alter it and redistribute
 * it freely, subject to the following restrictions:
 *
 * 1. The origin of this software must not be misrepresented; you must not
 *    claim that you wrote the original software. If you use this software
 *    in a product, an acknowledgment in the product documentation would be
 *    appreciated but is not required.
 *
 * 2. Altered source versions must be plainly marked as such, and must not be
 *    misrepresented as being the original software.
 *
 * 3. This notice may not be removed or altered from any source distribution.
 */

#if __GNUC__ >= 3 && __GNUC_MINOR__ >= 4 || __GNUC__ >= 4
#define	USED	__attribute__ ((used))
#elif defined __GNUC__
#define	USED	__attribute__ ((unused))
#else
#define	USED
#endif
#if defined (SUS)
static const char sccsid[] USED = "@(#)date_sus.sl	1.25 (gritter) 11/22/05";
#else
static const char sccsid[] USED = "@(#)date.sl	1.25 (gritter) 11/22/05";
#endif

#include	<unistd.h>
#include	<stdio.h>
#include	<string.h>
#include	<stdlib.h>
#include	<errno.h>
#include	<libgen.h>
#include	<time.h>
#include	<locale.h>
#include	<utmpx.h>
#include	<langinfo.h>

static unsigned	errcnt;			/* count of errors */
static int	bflag;			/* boot flag: don't write wtmp */
static int	uflag;			/* use UTC */
static char	*progname;		/* argv[0] to main() */
static time_t	now;			/* current time */

static void
usage(void)
{
	fprintf(stderr,
	"usage: %s [-u] [+format] [hhmm | mmddhhmm[[cc]yy]] [-a sss.fff]\n",
		progname);
	exit(1);
}

static void
badconv(void)
{
	fprintf(stderr, "%s: bad conversion\n", progname);
	exit(1);
}

static int
atot(char *s, int m)
{
	char *x;
	int i;

	i = (int)strtol(s, &x, 10);
	if (*x != '\0' || i > m || i < 0 || *s == '-' || *s == '+')
		return -1;
	return i;
}

static time_t
timeop(char *op)
{
	struct tm *tp;

	tp = uflag ? gmtime(&now) : localtime(&now);
	tp->tm_isdst = -1;
	tp->tm_sec = 0;
	switch(strlen(op)) {
	case 12:
		if ((tp->tm_year = atot(&op[8], 30000)) < 1970)
			goto badtime;
		tp->tm_year -= 1900;
		goto bypass;
	case 10:
		if ((tp->tm_year = atot(&op[8], 99)) < 0)
			goto badtime;
		if (tp->tm_year < 69)
			tp->tm_year += 100;
bypass:
		op[8] = '\0';
		/*FALLTHRU*/
	case 8:
		if ((tp->tm_min = atot(&op[6], 59)) < 0)
			goto badtime;
		op[6] = '\0';
		if ((tp->tm_hour = atot(&op[4], 23)) < 0)
			goto badtime;
		op[4] = '\0';
		if ((tp->tm_mday = atot(&op[2], 31)) <= 0)
			goto badtime;
		op[2] = '\0';
		if ((tp->tm_mon = atot(op, 12)) <= 0)
			goto badtime;
		tp->tm_mon--;
		break;
	case 4:
		if ((tp->tm_min = atot(&op[2], 59)) < 0)
			goto badtime;
		op[2] = '\0';
		if ((tp->tm_hour = atot(op, 23)) < 0)
			goto badtime;
		break;
	default:
		goto badtime;
	}
	return mktime(tp);
badtime:
	return (time_t)-1;
}

#if defined (__FreeBSD__) || defined (__NetBSD__) || defined (__OpenBSD__) || \
	defined (__DragonFly__)
static int
stime(time_t *t)
{
	struct timeval	tv;

	tv.tv_sec = *t;
	tv.tv_usec = 0;

	return settimeofday(&tv, NULL);
}
#endif	/* __FreeBSD__, __NetBSD__, __OpenBSD__, __DragonFly__ */

static void
settime(char *op)
{
	struct utmpx before, after;
	const char wtmpxfile[] = "/var/log/wtmp";
	time_t newtime;

	memset(&before, 0, sizeof before);
	memset(&after, 0, sizeof after);
	before.ut_type = OLD_TIME;
	after.ut_type = NEW_TIME;
	strcpy(before.ut_line, "old time");
	strcpy(after.ut_line, "new time");
	if ((newtime = timeop(op)) == (time_t)-1)
		badconv();
	gettimeofday(&before.ut_tv, NULL);
	if (stime(&newtime) < 0) {
		fprintf(stderr, "%s: no permission\n", progname);
		exit(1);
	}
	gettimeofday(&after.ut_tv, NULL);
#ifdef	__linux__
	system("/sbin/hwclock -w >/dev/null 2>&1");
#endif	/* __linux__ */
	if (bflag == 0) {
		pututxline(&before);
		pututxline(&after);
#if !defined (__hpux) && !defined (_AIX)
		updwtmpx(wtmpxfile, &before);
		updwtmpx(wtmpxfile, &after);
#endif	/* !__hpux, !__AIX */
	}
	exit(0);
}

static void
printtime(const char *cp)
{
	struct tm *tp;
	char fmt[256];
	register char *fp;
	char buf[256];
	int mod;
	char	*date_fmt;

	tp = uflag ? gmtime(&now) : localtime(&now);
	while (*cp != '\0') {
		if (*cp == '%') {
			const char	valid[] = "aAbBc"
#ifdef	SUS
					"C"
#endif	/* SUS */
					"dDeFgGhHIjmMnprRStTuUVwWxXyYZ%";
			fp = fmt;
			*fp++ = *cp++;
			if (*cp == 'E' || *cp == 'O')
				mod = *cp++;
			else
				mod = '\0';
			if (*cp == '\0') {
				putchar('%');
				goto end;
			}
			if (!strchr(valid, *cp)) {
				switch (*cp) {
				case 'N':
#ifndef	SUS
				case 'C':
#endif	/* !SUS */
#ifdef	_DATE_FMT
					if ((date_fmt = nl_langinfo(_DATE_FMT))
							== NULL ||
							*date_fmt == '\0')
#endif	/* _DATE_FMT */
						date_fmt =
						"%a %b %e %H:%M:%S %Z %Y";
					strcpy(fmt, date_fmt);
					break;
				case 'E':
				case 'O':
					goto next;
				case 'z': {
						  int hour, min;
#if defined (__FreeBSD__) || defined (__OpenBSD__) || defined (__DragonFly__)
						  long	timezone;
						  timezone = -tp->tm_gmtoff;
#endif	/* __FreeBSD__, __OpenBSD__, __DragonFly__ */

						  hour = -timezone / 3600;
						  min = (timezone / 60) % 60;
						  if (tp->tm_isdst > 0)
							  hour++;
						  fp = fmt;
						  if (hour <= 0) {
							  *fp++ = '-';
							  hour = -hour;
						  }
						  sprintf(fp, "%02d%02d",
								  hour, min);
					  }
					break;
				default:
					*fp++ = '%';
					if (mod)
						*fp++ = mod;
					*fp++ = *cp;
					*fp++ = '\0';
				}
			} else {
				if ((mod == 'E' && strchr("cCxXyY", *cp)) ||
						(mod == 'O' &&
						 strchr("deHImMSuUVwWy", *cp)))
					*fp++ = mod;
				*fp++ = *cp;
				*fp = '\0';
			}
			strftime(buf, sizeof buf, fmt, tp);
			fputs(buf, stdout);
		} else
			putchar(*cp);
next:
		cp++;
	}
end:
	putchar('\n');
}

static void
adjust(char *op)
{
	struct timeval tv;
	char *cp;

	if ((cp = strchr(op, '.')) != NULL) {
		if ((tv.tv_usec = atot(&cp[1], 999)) < 0)
			usage();
		tv.tv_usec *= 1000;
		*cp = '\0';
	}
	if ((tv.tv_sec = atot(*op == '-' ? &op[1] : op, 999)) < 0)
		usage();
	if (*op == '-') {
		tv.tv_sec = -tv.tv_sec;
		tv.tv_usec = -tv.tv_usec;
	}
	if (adjtime(&tv, 0) < 0) {
		fprintf(stderr, "%s: Failed to adjust date: %s\n", progname,
				strerror(errno));
		exit(1);
	}
	return;
}

int
main(int argc, char **argv)
{
	const char optstring[] = ":a:bu";
	char *adjustment = NULL;
	int i;

	time(&now);
	progname = basename(argv[0]);
	setlocale(LC_TIME, "");
#ifdef	__GLIBC__
	putenv("POSIXLY_CORRECT=1");
#endif
	while ((i = getopt(argc, argv, optstring)) != EOF) {
		switch (i) {
		case 'a':
			adjustment = optarg;
			break;
		case 'b':
			bflag = 1;
			break;
		case 'u':
			uflag = 1;
			putenv("TZ=GMT");
			tzset();
			break;
		case ':':
			usage();
			/*NOTREACHED*/
		default:
			/*fprintf(stderr, "%s: no TOY clock\n", progname);
			exit(2);*/
			badconv();
		}
	}
	if (adjustment) {
		if (uflag)
			badconv();
		adjust(adjustment);
	} else if (argv[optind] && argv[optind][0] == '+') {
		if (argv[optind][1] == '\0')
			goto dfl;
		printtime(&argv[optind][1]);
	} else if (argv[optind])
		settime(argv[optind]);
	else
	dfl:	printtime("%N");
	return errcnt;
}
